actors/migration/nv9/top.go

```diff
@@ -44,9 +44,21 @@ type Logger interface {
 	Log(level rt.LogLevel, msg string, args ...interface{})
 }
 
+// MigrationCache ByAddress calls need not be threadsafe
+// It must be safe to operation on any two output PreMigrationAddressCache concurrently
+type MigrationCache interface {
+	ByAddress(addr address.Address) MigrationAddressCache
+}
+
+// MigrationAddressCache need not be threadsafe
+type MigrationAddressCache interface {
+	Write(oldCid, newCid cid.Cid) error
+	Read(oldCid cid.Cid) (bool, cid.Cid, error)
+}
```

Comment:
I think this two-level approach causes unnecessary work (here, and later in Lotus), while not being fully general enough to solve the problem of cache key semantics.

That problem, to be clear, is that CID `Ba1` may be migrated to `Ba2` in one context (say as a cron queue entry) but to `Ba3` (or remain `Ba1`) in another (e.g. some piece of market actor state). We'll have this problem if the same data schema has two different interpretations (which is quite unlikely since the schema is ultimately just a heterogenous array). The two level arrangement solves that problem only if one of the contexts is top-level actor state. I don't think you've run into that problem with the simple caching behaviour currently in use, but we could do so later if we were unlucky, or implemented some kind of fully general cache interceptor across all state objects.

In this case I think the general solution is also simpler. That general solution is just an abstraction over a `map[[]byte]Cid`, where the context of a migration input can be encoded in the key. In the simplest case, both `Address` and `Cid` have cheap `[]byte` interpretations and can be used as keys directly. In more complex cases, path or context information can be prepended to the key.

Say we got unlucky and an AMT intermediate node has the same schema as some top-level actor type, and indeed could contain the same data but be migrated to different things in v3. To work around that, prefix keys with "actor/", "miner/sectors/" etc.

Coming back to the here and now, I don't think you need to do any of that key prefixing yet, but the current structure of representing context by the hierarchy of caches will be painful to extend if/when we do need more. And is just noisy now anyway. Instead, represent the context in hierarchy of the keys in a single cache.

The one thing you'll lose is that the cache will need to be synchronized. But that will be easy and unlikely to affect performance.

This will also make passing the cache around easier. There's just one, so it can be initialised as a member of `cachedMigrator` at the top, before the actor address is resolved.